Handle end lines in apply_text_edits

ndreas opened the pull request:
#12223 introduced a regression when the LSP is used along with `efm-langserver` and `diagnostic-languageserver`, maybe more. These servers send an end line that is larger than what nvim considers to be the end line, and that causes an `Index out of bounds` error with the change from #12223 that fetches the `end_bline`. This change adds tests for the error and adds checking to the values used.

tjdevries commented on test/functional/plugin/lsp_spec.lua:
```diff
@@ -831,6 +831,29 @@ describe('LSP', function()
         'å ä ɧ 汉语 ↥ 🤦 🦄';
       }, buf_lines(1))
     end)
+    it('applies edits with last line after vim\'s last line', function()
+      local edits = {
+        make_edit(0, 0, 5, 0, {"All replaced"});
+      }
+      exec_lua('vim.lsp.util.apply_text_edits(...)', edits, 1)
+      eq({
+        'All replaced';
+      }, buf_lines(1))
+      edits = {
+        make_edit(0, 0, 6, 0, {"All replaced"});
+      }
+      exec_lua('vim.lsp.util.apply_text_edits(...)', edits, 1)
+      eq({
+        'All replaced';
+      }, buf_lines(1))
+      edits = {
+        make_edit(0, 0, 6, 2, {"All replaced"});
+      }
+      exec_lua('vim.lsp.util.apply_text_edits(...)', edits, 1)
+      eq({
+        'All replaced';
+      }, buf_lines(1))
+    end)
```
It seems like each of these should be in their own little `it(...)` block. Otherwise they are changing the same buffer each time.

Each can have it's own little name why you're trying to do it `it('should apply edits to one extra row')`, `it('should apply edits to two extra rows')`, etc.

ndreas replied:
Fixed, and wrapped in a describe block to better give context

tjdevries commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -105,12 +105,20 @@ function M.apply_text_edits(text_edits, bufnr)
+      if #start_bline > 0 then
+        start_col = vim.str_byteindex(start_bline[1], start_col)
+      end
+    end
```
Perhaps this section could be wrapped up in another util function and shared between here and line 116 (since they're both doing the same thing). It seems easy to forget to handle these problems later (multi byte stuff, not enough lines, etc.). Should wrap them up in one spot.

ndreas replied:
Done!

tjdevries commented:
Two pretty small nitpicks. Thanks for fixing and adding tests! :D

ndreas committed:
Handle end lines in apply_text_edits

If the LSP sends an end line that is larger than what nvim considers to be the
last line, you get an Index out of bounds error when fetching the line from
nvim, a change that was introduced in #12223. This change removes the strict
indexing and checks the return value from nvim_buf_get_lines.

ndreas commented:
> Two pretty small nitpicks. Thanks for fixing and adding tests! :D

Not nitpicks, good things :)

clason commented:
Thanks for fixing this! Factoring out the conversion is also a good idea. I think this should be used in other utility functions as well -- could you look for other occurrences and adapt them? (`jump_to_location` comes to mind).

tjdevries commented:
Maybe I'll merge in this fix first and then if you want @ndreas you can use the funciton you just made in some of the other places @clason mentioned. Or would you prefer to do it in this PR? 

I'm just thinking it is good to fix this problem if it's breaking some servers.

ndreas commented:
I'd prefer it if this is merged since it breaks my nix and shell formatting :) I can fix the other places in another PR, happy to contribute!

tjdevries commented:
Awesome :) feel free to open up another PR if you want to improve more of this area. Nice work and thanks for adding the tests!
